pkg/webrtcclient/client_test.go

```diff
@@ -0,0 +1,153 @@
+
+package webrtcclient
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/pion/webrtc/v3"
+)
+
+// An example offer sdp string with a single data channel
+const fakeSDP string = `v=0
+o=- 7723759704302471067 2 IN IP4 127.0.0.1
+s=-
+t=0 0
+a=group:BUNDLE 0
+a=extmap-allow-mixed
+a=msid-semantic: WMS
+m=application 9 UDP/DTLS/SCTP webrtc-datachannel
+c=IN IP4 0.0.0.0
+a=ice-ufrag:yS7P
+a=ice-pwd:DML7anbSv0d83sY1TspIUSAM
+a=ice-options:trickle
+a=fingerprint:sha-256 AC:2D:41:40:67:D7:AB:B8:42:51:A6:E8:A3:2B:AE:26:24:76:FA:E6:8C:25:78:FF:2B:69:F3:16:DC:E3:BB:3A
+a=setup:actpass
+a=mid:0
+a=sctp-port:5000
+a=max-message-size:262144
+`
+
+type TestObserver struct {
+	dc               *webrtc.DataChannel
+	onADBDataChannel func(*webrtc.DataChannel)
+	onError          func(error)
+}
+
+func (o *TestObserver) OnADBDataChannel(dc *webrtc.DataChannel) {
+	if o.onADBDataChannel != nil {
+		o.onADBDataChannel(dc)
+	}
+}
+
+func (o *TestObserver) OnError(err error) {
+	if o.onError != nil {
+		o.onError(err)
+	}
+}
+
+func recvWithTimeOut(sendCh chan interface{}, msg interface{}) error {
+	select {
+	case m := <-sendCh:
+		// Reshape instead of type check here: the client is allowed to send whatever
+		// type it wants as long as has the expected fields.
+		return Reshape(m, msg)
+	case <-time.After(1 * time.Second):
+		return fmt.Errorf("time out")
+	}
+}
+
+func TestClientSendsInitialMessage(t *testing.T) {
+	observer := TestObserver{}
+	signaling := Signaling{
+		SendCh: make(chan interface{}),
+		RecvCh: make(chan map[string]interface{}),
+	}
+	// This won't return, so run it in another routine
+	go func() {
+		_, _ = NewDeviceConnection(&signaling, &observer)
+	}()
+
+	var msg RequestOfferMsg
+	if err := recvWithTimeOut(signaling.SendCh, &msg); err != nil {
+		t.Fatalf("Client didn't send the request-offer message in time: %v", err)
+	}
+	if msg.Type != RequestOfferMsgType {
+		t.Fatalf("First message has wrong type: '%v', expected '%s'", msg.Type, RequestOfferMsgType)
+	}
+
+}
+
+func TestClientGeneratesAnswer(t *testing.T) {
+	observer := TestObserver{}
+	signaling := Signaling{
+		SendCh: make(chan interface{}),
+		RecvCh: make(chan map[string]interface{}),
+	}
+	// This won't return, so run it in another routine
+	go func() {
+		_, _ = NewDeviceConnection(&signaling, &observer)
+	}()
+	// The first message is always the request-offer
+	var _offerMsg interface{}
+	if err := recvWithTimeOut(signaling.SendCh, &_offerMsg); err != nil {
+		t.Skipf("Client didn't send request-offer in time: %v", err)
+	}
+	signaling.RecvCh <- map[string]interface{}{
+		"type": "offer",
+		"sdp":  fakeSDP,
+	}
+	var msg webrtc.SessionDescription
+	if err := recvWithTimeOut(signaling.SendCh, &msg); err != nil {
+		t.Fatalf("Client didn't send the answer message in time: %v", err)
+	}
+	if msg.Type != webrtc.SDPTypeAnswer {
+		t.Fatalf("Expected 'answer' message, but got '%v' instead", msg.Type)
+	}
+}
+
+func TestClientGeneratesICECandidate(t *testing.T) {
```

Comment:
nit: I find it useful to lay out the test methods into 3 clear blocks: arrange, act, and assert separated by new line.

Reply:
Done